Setup pipelines (#6)
apollographql/apollo-cli: Contributing.md

### Responding to issues

In addition to reporting issues, a great way to contribute to Apollo is to respond to other peoples' issues and try to identify the problem or help them work around it. If you’re interested in taking a more active role in this process, please go ahead and respond to issues. And don't forget to say "Hi" in the [Apollo Spectrum community](https://spectrum.chat/apollo)!

### Small bug fixes

For a small bug fix change (less than ~20 lines of code changed), feel free to open a pull request. We’ll try to merge it as fast as possible and ideally publish a new release on the same day. The only requirement is, make sure you also add a test that verifies the bug you are trying to fix.

### Suggesting features

Most of the features in Apollo Client came from suggestions by you, the community! We welcome any ideas about how to make Apollo  better for your use case. Open up a new feature request / discussion issue with your details.

> **Note**: Feature requests and non-bug related discussions are no longer managed in this repo's issue tracker. Feature request and/or discussions opened here will be closed.

Reviewer: If we drop the mention of the feature request repo, this should go too :)

## Big Changes or New Features

For significant changes to a repository, it’s important to settle on a design before starting on the implementation. This way, we can make sure that major improvements get the care and attention they deserve. Since big changes can be risky and might not always get merged, it’s good to reduce the amount of possible wasted effort by agreeing on an implementation design/plan first.

> 1. **Open an issue.** Open an issue about your bug in this repo, or about your feature request

Reviewer: ```suggestion
1. **Open an issue.** Open an issue about your bug or feature request in this repo. 
```

## Developing the CLI

### Quick Start

If you have all of the dependencies setup and are ready to get right into contributing, run the following command after cloneing the repo:

```
cargo run -- help
```

This step will install of the dependencies and print out the help command from the CLI. To run develop commands and test them out locally, you can use the `cargo run` command and pass commands and headers like this:

```
cargo run -- <command name> --<flag>
```

Cargo will proxy your commands and flags to the built CLI project.

### Deep dive

**Prerequisites:**
This project is written in Rust and setup as a multi crate (aka package) project to allow us to share common code across multiple packages.  To get started you need to have the Rust toolchain installed. The best way to do this is to visit [the amazing rust site](https://www.rust-lang.org/learn/get-started) and follow their install instructions.

To ensure you are ready, you should be able to run this:

`cargo --version` 

and get something back that looks like this:

`cargo 1.42.0 (86334295e 2020-01-31)`

**Testing the project**

> Rust has great built-in test tooling. You can write tests in-line or under the tests folder for each crate. Tests under the `tests` folder should be integration whereas inline should be unit tests (as I understand from the conventions). You can run `cargo test` to test all of the crates at once.

Reviewer: ```suggestion
Rust has great built-in test tooling. You can write tests in-line or under the tests folder for each crate. Tests under the `tests` folder should be integration whereas inline should be unit tests. You can run `cargo test` to test all of the crates at once. For more info on how to write tests, look to similar parts of the codebase or read the [rustlang.org article](https://doc.rust-lang.org/book/ch11-00-testing.html) on testing
```

> If you have your editor setup with the Rust extension, you can click the codelens “Test” command to run individual tests right from you editor! How cool is that?!

**Opening a Pull Request**
Once you have fixed a bug, created the next great feature, or improved the project in some other way it is time to open a pull request! To learn more about how to do this, follow [this great guide](https://opensource.guide/how-to-contribute/#opening-a-pull-request). When you open a PR, the pipeline steps will run to ensure the changes will work for users of this library. To learn more about the pipeline, keep reading below!

## Pipelines

This project uses GitHub Actions to run a continuous integration and delivery pipeline. Every code change will be run against a few steps to help keep the project running at its peak ability

> * **CLA Check**: If you haven’t signed the Apollo CLA, a bot will comment on your PR asking you to do this

Reviewer: Should we link to the that here? Small improvement that could definitely be skipped

Author: I still need to turn it on haha

* **Tests**: The CI will run the `cargo test` command across there different architectures (Mac OS, Linux, and Windows). If your build fails on a platform that can’t test on don’t worry! The team will be able to help you out as we run all three platforms to make sure everyone has a great experience.

> * **Coverage**: This project runs a job to collect information on how much of the library has been tested using code coverage tools. This is a WIP but you may see a status check related to how you have improved (or lowered 😢) the amount covered. Don’t worry, it happens to all of us and we are hear to help out!

Reviewer: ```suggestion
* **Coverage**: This project runs a job to collect information on how much of the library has been tested using code coverage tools. This is a WIP but you may see a status check related to how you have improved (or lowered 😢) the amount covered. Don’t worry, it happens to all of us and we are here to help out!
```
